### DISCUSSION TITLE
[RFC] 关于 AI Native 模块的建设

### DISCUSSION THREAD

**Discussion by Ricbet** [opensumi/core#3250] [RFC]
## 介绍
打造 OpenSumi 内置的 AI 原生模块，皆在提供 IDE 内各个功能模块（如编码、搜索、调试、终端、Command等）与 AI 相结合的可能性。

我们打造 AI Native 模块的目标主要集中在几个方面

- 帮助企业或个人快速打造大模型时代下的 IDE 产品（IDE Copilot）
- 将 AI 原生的基础辅助能力融入到 IDE 用户核心的工作流当中
- 帮助 AI & IDE 相结合的想象力提供可落地的场景
- 跨端的 AI 集成能力（本地客户端、云端、纯前端）

> 注: 对于框架来说，AI Native 模块提供的是 AI 基础集成能力，本身不提供模型服务（如 OpenAI）
## 功能
具体要做的事情包括但不限于以下内容

- 内置的 AI Chat 助手，并提供 Agent 开放能力与快捷指令能力的注册和扩展
- Inline Chat 能力，在编辑器内通过丰富的交互与 AI 能力来生成或理解代码
- 内置代码补全，包含块补全、行内补全等基础能力
- 运行报错诊断，提供程序运行错误或静态语法问题时的检测能力
- 提供全新的 OpenSumi Design UI 视图
> 初版思维导图，不代表最终方案

![image](https://github.com/opensumi/core/assets/20262815/08b41c76-dbaa-4720-8441-9077ca93871a)

> 初版 UI，不代表最终方案

![image](https://github.com/opensumi/core/assets/20262815/497a7588-d736-4ac6-bd8b-a0adf29eff85)

## TODO LIST
> 以下事项是基于 chore/ai-native-module 分支的基础之上

#### AI Chat 助手

- [ ] 基础对话能力
- [ ] 流式渲染能力
- [ ] 提供模型服务的接口接入能力
- [ ] 支持自定义模版渲染
- [ ] 输入框支持快捷指令注册
- [ ] 支持 agent 能力注册
- [ ] 代码渲染使用 monaco editor（待定）
	##### 关于 API

  - [ ] 提供模块 API
  - [ ] 提供插件 API（由于 vscode 相关插件 api 仍处于 proposed 阶段，暂时不考虑对齐 vscode 的插件 api）
#### Inline Chat 能力

- [ ] 提供按钮注册能力
- [ ] 提供输入框 Chat 的能力
- [ ] 输入框快捷指令注册（如 "/fix xxxx"等）
  ##### 根据模型接口的 answer 使用不同的预览策略来渲染

  - [ ] 内嵌的 diff editor
  - [ ] 内嵌的 code editor
  - [ ] 流式输出渲染

#### 问题诊断

- [ ] 问题面板的智能诊断

#### monaco editor 版本跟进

- [ ] 升级 monaco editor 版本。inline chat 想做的很多事情都是需要依赖最新版本的 monaco editor

#### 全新的 OpenSumi Design UI 视图

- [ ] 从 AI Native 模块当中抽离出 UI Design，现在是耦合在一起的
- [ ] 适配白色主题

> 待续...

## 接下来
AI Native 模块的建设最早从九月底就已经开始投入，相关 PR 见 [#3077](https://github.com/opensumi/core/pull/3077)
不过一直处于 WIP 的状态，前期也是在不断的摸索与尝试，发布模式也一直脱离于主干分支
现在我们认为它已经是一个较为稳定的版本，计划将该 chore/ai-native-module 分支的代码分批次的合入主分支中，并重构里面的代码内容，保证健康的拓展性与可维护性

我们会陆陆续续开始这方面的工作

**Comment by Ricbet** [comment]
## 议题
1. OpenSumi Design 拆分
2. 具体功能拆分
   1. 内置 Chat 面板
   2. inline chat
   3. 代码补全
   4. 终端运行识别与诊断
3. 模块 API 和插件 API 的设计
4. 相关 Config 配置设计
5. 代码文件结构优化

### OpenSumi Design 拆分
**问题所在:**

1. Design 与 AI Native 模块强耦合，架构设计不合理，应该独立拆开
2. 主题不好适配。里面存在很多我们自定义的 color token，两款默认主题比较好适配，但其他主题的适配情况就比较糟糕

**重构策略**

有两种重构方案

1. 设置一个 AppConfig 的配置项（如：useSumiDesign），然后在各个模块内做判断，来渲染不同的组件样式
   1. 好处：简单方便
   2. 坏处：无法处理更为复杂的定制化需求，而且后续如果 OpenSumi Design 有一个迭代升级就不太好处理
2. 单独抽离成 design 模块（或者放在 core-browser），在这个模块内去做样式组件设计，包括渲染各种 slot render 以及主题色注册等。然后在其他模块里可以直接使用里面的 design 组件等（例如 editor tab）
   1. 好处：扩展性强，定制化高，也好做升级
   2. 坏处：较为复杂

> 主题适配问题暂时没想到好的解决办法


### 具体功能拆分

- **Chat 面板：** 保留现状，优化代码结构即可
- **inline chat：**
   - 当前版本的 inline chat 与 vscode 的 inline chat 有很大的区别，不过阿里集团侧在内部已经实现了后者。这块可以考虑同时兼容两种 inline chat 的模式
   - sumi inline chat 功能还比较简陋，比如没有 input 输入框，也不是流式输出。可以考虑先将流式输出给做了，input 输入框需要设计师来参与再开始搞
   - **现阶段可以先将 vscode 的 inline chat 给内置实现了，然后由模块 API 去控制选择哪一种模式**
- **代码补全：**
   - 当前第一版的代码补全实现比较简单，只包含当前文件的上下文信息，后续这块需要重新规划调整
   - 现阶段可以保留现状，直接移植过去即可
- **终端运行识别与诊断：**
   - 将蚂蚁内部已经实现了的 terminal decoration 下沉到 opensumi 里，实现类似下面这样的效果

![image](https://github.com/opensumi/core/assets/20262815/23f33ef0-54cc-41ca-b7b1-7f42628be8de)

### 模块 API 和插件 API 的设计
#### 模块 API
主要是给集成方使用，通过实现 AiNativeCoreContribution 接口的方式来调用模块 API
![image](https://github.com/opensumi/core/assets/20262815/82e9f647-6275-4b3f-b7c1-403484e2f1a4)

比如 inline chat api（这三个按钮都是上层集成方注册的）
![image](https://github.com/opensumi/core/assets/20262815/2347f261-e2dd-426d-ad8c-d276d72f346b)

当前设计的模块 API 有这些，只涉及到 run（顶部的 run 按钮）、inline chat 以及智能补全
```typescript
export interface AiNativeCoreContribution {
  /**
   * 注册 ai run 的能力
   * @param registry
   */
  registerRunFeature?(registry: IAiRunFeatureRegistry): void;
  /**
   * 注册 inline chat
   * @param registry
   */
  registerInlineChatFeature?(registry: IInlineChatFeatureRegistry): void;
  /**
   * 通过中间件扩展部分 ai 能力
   */
  middleware?: IAiMiddleware;
}
```
其中 middleware 字段主要是用于代理 monaco 的 api，以达到扩展的目前，当前仅可扩展智能补全 API
```typescript
export type IProvideInlineCompletionsSignature = (
  this: void,
  model: monaco.editor.ITextModel,
  position: monaco.Position,
  token: CancellationToken,
  next: (reqBean: CompletionRequestBean) => MaybePromise<CompletionResultModel | null>,
  completionRequestBean: CompletionRequestBean,
) => MaybePromise<CompletionResultModel | null>;

export interface IAiMiddleware {
  language?: {
    provideInlineCompletions?: IProvideInlineCompletionsSignature;
  };
}
```


目前遇到的需求只涉及到以上这三个，后续如果有新的需求需要实现，**直接扩展 AiNativeCoreContribution 接口即可**

#### 关于 AiBackService
目前所有的 ai 接口请求都是用的 AiBackSerivcePath 这个 DI，位于 Node 层
![image](https://github.com/opensumi/core/assets/20262815/c3c80e01-573b-4d99-98ca-23d3f8b2c609)
在 opensumi 内部，ai native 模块都是直接使用 AiBackSerivcePath 这个 DI 去调用 AiBackService，默认是一个空实现
```typescript
@Injectable()
export class BaseAiBackService
  extends RPCService<IRPCGptService>
  implements IAiBackService<IAiBackServiceResponse, Readable>
{
  async request<T = IAiBackServiceResponse<string>>(
    input: string,
    options: IAiBackServiceOption,
    cancelToken?: CancellationToken,
  ): Promise<T> {
    return void 0 as T;
  }
  async requestStream<T = IAiBackServiceResponse<string>>(
    input: string,
    options: IAiBackServiceOption,
    cancelToken?: CancellationToken,
  ): Promise<T> {
    return void 0 as T;
  }

  async requestCompletion<T = CompletionResultModel>(input: IAiCompletionOption, cancelToken?: CancellationToken) {
    return void 0 as T;
  }

  async reportCompletion<T = IAiReportCompletionOption>(input: IAiReportCompletionOption) {}

  async destroyStreamRequest(sessionId: string) {}
}
```
而集成方可以用 AiBackSerivceToken 这个 token 去注册一个 AiBackService 来替代这个默认实现，进而可以将所有的 ai 接口请求都进到自己的 service
```typescript
{
  token: AiBackSerivceToken,
  useClass: OverrideAiBackService
}
```

这样的设计有助于其他的模块如果仅仅只是想调用 AI 接口服务的时候，从 core-common 模块使用这个 DI 即可，而不必使用整个 AI Native 模块

比如 AI 智能解决冲突功能
![image](https://github.com/opensumi/core/assets/20262815/4db6df1a-493f-42e9-9fdf-864a38ac3afe)

#### 插件 API
目前只实现了 chatAgent 相关的 proposed 插件 API，已经具备了 agent 能力
![image](https://github.com/opensumi/core/assets/20262815/ead1578e-b112-4305-bcc3-ea9f9fee1561)

其他 API 暂未开始<br />目前对于 AI Native 来说其他 API 优先级还没那么高，且仍处于 proposed 阶段，现在跟进的话会有一定的成本
### 相关 Config 配置设计
目前已有的配置项如下
```typescript
export interface IAiNativeCapabilities {
  /**
   * 是否使用 opensumi design UI 风格
   */
  supportsOpenSumiDesign?: boolean;
  /**
   * 问题面板是否使用 ai 能力
   */
  supportsAiMarkers?: boolean;
  /**
   * 是否使用 Chat 助手能力
   */
  supportsAiChatAssistant?: boolean;
  /**
   * 是否使用 inline chat 能力
   */
  supportsInlineChat?: boolean;
  /**
   * 是否使用代码智能补全能力
   */
  supportsInlineCompletion?: boolean;
  /**
   * 是否使用 ai 智能解决冲突的能力
   */
  supportsConflictResolve?: boolean;
}

export interface AiNativeConfig {
  capabilities?: IAiNativeCapabilities;
}
```
后续等终端的 decoration 的设计内置之后，可再新增一个 capabilities 配置

### 代码文件结构优化
![image](https://github.com/opensumi/core/assets/20262815/4710d676-56af-415d-8e56-205131de7bca)

- ai-project：智能生成项目功能，目前主要是用来主动给 Chat 面板发送消息（例如展示生成的文件信息等），由于已经实现了插件的 Chat API，可以将这个功能移植到插件里去
- ai-sumi：用于在 Chat 面板实现 /IDE 这样的指令，能通过自然语言调用 command 命令，建议保留
- components：部分共用组件可移植到 core-browser 中（例如 sumi inline chat 组件在 terminal 或 3-way 里都需要用到）
- contextkey：保留
- diff-widget：在编辑器里展示 diff 代码，主要用于 inline chat 的生成结果展示。保留
- inline-chat-widget：sumi inline chat 的实现
- inline-completions：智能代码补全的实现
- merge-conflict：传统模式下的智能解决冲突实现
- model：管理流式输出的 manager
- override：opensumi design 的实现，覆写了很多样式主题以及 slot render
- prompts：prompt 管理，目前仅在 ai-sumi 里用到，可移植到 ai-sumi 里
- run：顶部 run 按钮的实现

后续要将所有 chat 相关的代码放在 chat 目录里

**Comment by Ricbet** [comment]
todolist

- [x] OpenSumi Design 拆分 https://github.com/opensumi/core/pull/3350 
- [x] inline chat 实现 https://github.com/opensumi/core/pull/3378
- [x] chat 面板相关功能 https://github.com/opensumi/core/pull/3394
- [x] 内置代码补全 by @bytemain  https://github.com/opensumi/core/pull/3400
- [x] 智能解决冲突 3way 交互 @Ricbet https://github.com/opensumi/core/pull/3404
- [x] 智能解决冲突传统交互 @Ricbet 
- [x] chat 面板支持自定义视图 @Ricbet  https://github.com/opensumi/core/pull/3499
- [x] terminal inline chat api @Ricbet https://github.com/opensumi/core/pull/3427
- [x] 智能终端功能 @life2015  https://github.com/opensumi/core/pull/3422

**Comment by xinrulianhua** [comment]
可以单独抽离成AI Copilot ，支持与monaco editor交互？
